core/shared/src/main/scala/fs2/internal/Scope.scala

```diff
@@ -0,0 +1,295 @@
+    *
+    */
+  def acquire: F[Option[F[Unit]]] = {
+    F.map(state.modify { s =>
+      if (s.notOpen_?) s
+      else s.copy(refCount = s.refCount + 1)
+    }) { c =>
+      if (c.previous.notOpen_?) None
+      else Some(cancelAcquire())
+    }
+  }
+
+
+
+}
+
+object Scope {
+  def newRoot[F[_]: Sync]: Scope[F] = new Scope[F](new Token(), None)
+
+  /**
+    * State of the scope
+    *
+    * @param closed             Indicates whether the scope is closed. Empty, when scope is open, that means resources
+    *                           are about to e acquired (see refCount) or they are being used.
+    *                           When Some(true) that indicates scope is fully closed and resources were released. Note that
+    *                           they may not have even at that time to be fully finalized, but all users of that resources
+    *                           shall fully release them at this point.
+    *
+    * @param refCount           Contains a reference count which indicates that scope's resources may not start their release
+    *                           sequence until this is eq 0.
+    *
+    *                           When resource is asynchronously constructed this increments, to indicate that resource are being acquired
+    *                           but not yet finished their acquisition phase. The resource will decrement this number when either
+    *                             (a) the resources finishes its acquisition before scope is closed
+    *                             (b) if the acquisition of the resource was interrupted during the scope being closed before that acquisition
+    *                                 finished.
+    *
+    *
+    * @param refReleased        A callback  to be invoked, when this close was closed during asynchronous execution or
+    *                           before all linked `remote` scopes to this scope were unlinked.
+    *                           This must be invoked, when defined and when refCount is decrementing from 0 to 1 and only at that time.
+    *
+    * @param resources          All acquired resources (that means synchronously, or the ones acquired asynchronously) are
+    *                           registered here. Note that the resources are released in reverse order as they were acquired.
+    *
+    * @param children           Children of this scope. Children may appear during the parallel pulls where one scope may
+    *                           split to multiple asynchronously acauired scopes and resources. Still, likewise for resources they are released in reverse order.
+    *
+    *
+    * @tparam F
+    */
+  final private[Scope] case class ScopeState[F[_]](
+    closed: Option[Boolean]
+    , refCount: Int
+    , refReleased: Option[F[Unit]]
+    , resources: ListMap[Token, Resource[F]]
+    , children: ListMap[Token, Scope[F]]
+  ) {
+
+    def open_? : Boolean = closed.isEmpty
```

Comment:
Style thing: we don't use the `_?` syntax anywhere else in the project so I'd prefer to call these `isOpen`, `isClosed`, `isClosing`, `isClosed`

Reply:
ok